### DISCUSSION TITLE
Restructuring how we store resource links

### DISCUSSION THREAD

**Discussion by Bhupesh-V** [developersIndia/resources#21] [Ideas]
Right now everything is messy
- one big file, ughh!
- people add tips, advices everywhere
- sometime comments on resource are also present

Need an automated way to cleanly manage all resource without the need of a website. 

All suggestions are welcome 🙏🏽

**Comment by Bhupesh-V** [comment]
A draft solution would be to have separate folders based on common topics like `languages`, `cs`

each folder will have a `links.json` looking like this

```json
{
    "links":[
        {
            "title": "dksjklds",
            "url": "https://url.com",
            "type": "book",
           "level": "beginner",
            "reviews": [
                {
                    "comment": "Costly, But Worth the money!"
                }
            ]
        },
        {
            "title": "dksjklds",
            "url": "https://url.com",
            "category": "video",
            "reviews": [
                {
                    "comment": "Costly, But Worth the money!"
                }
            ]
        }

    ]
}
```

**List of resource types**:
- Book
- Video
- Cheatsheet
- GitHub
- Blog
- Course

a different json file called `tips.json` can be used to store advice, agreed by the community or even just one-off opinions

```
{
    "tips": [
        {
            "comment": "",
            "user": "",
            "user_profile_link": ""
        }
    ]
}
```

We use these json files to automatically create the well structured README.
[Here is a sample script](https://github.com/Bhupesh-V/til/blob/master/build.py) that can be used as an inspiration 

### Pros
- Contributions would be easier and streamlined. since different categories and folders
- this will help us automate the task of importing resources from discord as well

### Cons
- Bit of over-engineered

**Reply by Bhupesh-V** [reply]
here is how a sample README might look like
https://gist.github.com/Bhupesh-V/2fe59be3fb1872cb0a2ddc408958c142

**Reply by jarmos-san** [reply]
Agreed & +1 for structuring the topics into separate folders.

I think it's best to start small & gradually build from there, so it would be best to drop the `tips.json` idea temporarily. It shouldn't be a big deal to come back to it with a more streamlined & polished experienced. On top of that, I guess there should be an automatic way to pull the community pinned messages from Discord to this wiki.

**Reply by BlackGoku36** [reply]
> A draft solution would be to have separate folders based on common topics like languages, cs

Agreed

> Bit of over-engineered

Agreed Too

How about we have different files for categories and tabulate resources and have tips section beneath? like:

In ComputerGraphics.md:
<table>
  <tr>
    <th>Title/Link</th>
    <th>Category</th>
 <th>Language</th>
    <th>Comment</th>
  </tr>
  <tr>
    <td><a href="https://www.scratchapixel.com/index.php?redirect">Scratchapixel</a></td>
    <td>Article</td>
 <td>C++</td>
    <td>Theoretical resources with code</td>
  </tr>
  <tr>
    <td><a href="https://www.youtube.com/user/shiffman">Daneil Shiffman</a></td>
    <td>Videos</td>
   <td>JS (P5)</td>
    <td>Creative coding</td>
  </tr>
  <tr>
    <td><a href="https://github.com/Kode/Kha">Kha</a></td>
    <td>Tool</td>
   <td>Haxe</td>
    <td>Open source, cross-platform multimedia framework</td>
  </tr>
</table>

Tips/Advice:

- Level up by writing your own project

- Use graphics abstraction instead of writing bare Metal/OpenGL/Vulkan/DX12 code

We can enforce arranging columns in a particular order (Like articles come first, then come videos, and than tools. Also sub-groups are arranged in alphabetical order)

**Reply by jarmos-san** [reply]
@BlackGoku36 yeah you're right agreed with tabularizing the resources, doing that shouldn't be an issue. But serializing the resources for maintainability & portability is the important task at hand right now.

For example, the `tips.json` file can be used easily to give due credit to whoever suggested something helpful. On top of that, once the resources are properly serialized in the first place, automation should be a piece of cake. And the respective README files in each folder can be presented in whatever way as needed.

So, like right now, the priority should be, to think of a way to store the resources in a JSON (or YAML?) file for reasons mentioned above. And how exactly should we proceed to achieve this task is the first step.

**Reply by Bhupesh-V** [reply]
Cool thanks a lot for the suggestions @Jarmos-san @BlackGoku36 👍🏽 
So now the follow up things to decide on are:

- [ ] What categories we need ? (will start a separate thread for this)
- [ ] How the build process looks like (will use Github Actions to run a cron) ?
- [ ] What other things need to be considered before we start with a bot/service to  link the the resources channel to this repo

**Reply by BlackGoku36** [reply]
> What other things need to be considered before we start with a bot/service to link the the resources channel to this repo

For that, we could have bot in resource channel that with command create pull request?
For example, command could be something like: 
```
append {
category: "Computer Science",
subcategory: "Computer Graphics",
title: "Something",
link: "https://somelink.com",
type: "Article",
language: "C",
comment: "Lalalalalala"
}
```
or whatever json schema we will use.
Bot than can ask for account or whatever detail that can be use to credit the contributor. And finally ask whoever responsible for accepting this to accept it (maybe with simple emoji reaction).
Resource channel could only have a resource list which is fetched from the repo.
So, it will be something like this:
Contributor enters resources -> bot ask for review -> bot push to repo -> bot fetch new repo details and update the resource list in resources channel.

**Reply by Bhupesh-V** [reply]
I like this, I had a similar idea a while back 👌🏽 

- we can also use the concept where if a minimum amount of reactions are done on a message, the PR will be generated (this emphasises community involvement)
- Something to figure out do we actually need to credit the person who posted, some people might not like there name/profile being publicly shared

> Contributor enters resources -> bot ask for review -> bot push to repo -> bot fetch new repo details and update the resource list in resources channel.

Seems like a good idea 🔥 , will refine this when we start working on the bot

**Comment by Bhupesh-V** [comment]
### Thread for categories to include

All these categories will be folders, followed by JSON files for each. E.g `cpp.json`, `javascript.json`
1. Langauges
   - C
   - C++
   - JavaScript (will include frameworks)
   - Literally any other language.....
2. Computer Science
    - DSA
    - Computer Graphics
    - Machine Learning
3. Software Engineering
    - Front-end
    - Back-end
    - DevOps
    - Interview Prep
4. Miscellaneous




_Reply below for suggestions_

**Reply by jarmos-san** [reply]
Wonderful! Excited to see the project is gradually taking shape. But I've a couple more suggestions which should make maintaining the project much more easier.

### Suggestions for Project Maintenance

1.  Edit the above thread as you receive suggestions for more topics so that it acts as a one-stop solution for readers. GitHub tends to collapse long threads so that subtle task should be rather helpful to many.
2. Another suggestion I've is to create a draft [JSON schema](http://json-schema.org/) first. This should take some clean-up burden off of from the maintainer's shoulders. That way contributors will no longer be able to add any arbitrary properties to the respective JSON file(s).

### Additional Categories to Include

That said, here are a few more categories I can think of right away on top of my head. Will update this thread & leave a ping in the server for a notification (or reminder) of the same.

5. Interview Prep
6. Freelancing

**NOTE**: Strongly suggest drafting an initial JSON schema which will take off a huge burden of maintaining the JSON files.

**Reply by Bhupesh-V** [reply]
> Another suggestion I've is to create a draft JSON schema first. This should take some clean-up burden off of from the maintainer's shoulders. That way contributors will no longer be able to add any arbitrary properties to the respective JSON file(s).

TIL about JSON schema, will incorporate this for sure. Will spend some time testing the Python client over the weekend

>Interview Prep
Freelancing

Interview Prep comes under SE, what about Freelance, a new category altogether
Are we going to have sub-categories inside this as well 👀 ?

**Reply by jarmos-san** [reply]
> ...what about Freelance, a new category altogether

I guess it should come under a separate topic (maybe Career Advice)?

**Comment by Bhupesh-V** [comment]
Hey, we have finally re-structured the way we store resource links. The JSON schema is available here https://github.com/developersIndia/resources/blob/master/resource.schema

Closing this for now. Feel free to open up a new discussion for further development & questions
